### PR TITLE
remove redundant volatile

### LINKED ISSUE #328: remove redundant volatile
### Your question

There are some `volatile` variables in the project, but all of its assignment methods use the `synchronized` keyword.

### Your scenes

> com.alipay.sofa.registry.common.model.store.Watcher#pushedVersion

image

```java
  public synchronized boolean updatePushedVersion(long v) {
    if (pushedVersion < v) {
      this.pushedVersion = v;
      return true;
    }
    return false;
  }
```

### Your advice

For such variables, the volatile modifier can be removed.

### Environment

- SOFARegistry version:
- JVM version (e.g. `java -version`):
- OS version (e.g. `uname -a`):
- Maven version:
- IDE version:

### PR DESCRIPTION
**Pull request by ashlee618**
### Motivation:

Remove redundant volatile modifiers.

Because their assignment methods are all qualified with synsynchronized, their visibility is guaranteed.

### Modification:

Describe the idea and modifications you've done.

### Result:

Fixes #328.

### CODE REVIEW

**Comment by NickNYU** [comment]
Thanks for commit, here's what I thought why it needs to be `volatile`d:

we leverage on the `leaderInfo` for a leader comparasion and validation, this code block is not proteced by `synchronized`, that's why we still need a `volatile` here, but other maps, yes, maybe volatile is redundant.

```java
@Override
  public boolean amILeader() {
    return amILeader(leaderInfo.leader);
  }
  protected boolean amILeader(String leader) {
    return StringUtil.equals(myself(), leader) && leaderNotExpired();
  }
```

**Comment by ashlee618** [comment]
`synchronized` 修饰了 `leaderInfo` 唯一的赋值方法，他的可见性应该得到保证了，这里只是对 `leaderinfo` 进行读操作，感觉应该没有可见性的问题。

```java
  public void elect() {
    synchronized (this) {
      if (isObserver) { // 如果是Observer，不参与选主
        leaderInfo = doQuery();
        LOG.info("meta role: Observer, leaderInfo: {}", leaderInfo);
      } else {
        leaderInfo = doElect();
      }

      if (amILeader()) {
        onIamLeader();
      } else { // 我不是leader
        onIamNotLeader();
      }
    }
  }
```

还是说这里有其他特殊的场景（重排序？），可以再详细解答一下嘛。

**Comment by NickNYU** [comment]
> Is it because of `synchronized (this)` rather than `synchronized (Class)` here?
> 
> It seems necessary to add `volatile` here.
> 
> ```java
>   public void elect() {
>     synchronized (this) {
>       if (isObserver) { // 如果是Observer，不参与选主
>         leaderInfo = doQuery();
>         LOG.info("meta role: Observer, leaderInfo: {}", leaderInfo);
>       } else {
>         leaderInfo = doElect();
>       }
> 
>       if (amILeader()) {
>         onIamLeader();
>       } else { // 我不是leader
>         onIamNotLeader();
>       }
>     }
>   }
> ```

It is because the `leaderInfo` may be accessed by other threads while it's modifying simultanously.

And the `getter` method is not `synchronized`, which means, it's not protected by we call `store_store` or `store_load` mechanism, you could refer this knowledge to JVM locks.

By leveraging on `volatiled` keyword, it will trigger a `store_load` actively, the below code comes from newest OpenJDK, might be slightly diff from JDK-8 or others, but it doesn't matter, because I want to show you the way, how Jvm is handling with `volatile`:
BTW, this is for `x86` specified:
```c++
instruct membar_volatile(rFlagsReg cr) %{
  match(MemBarVolatile);
  effect(KILL cr);
  ins_cost(400);

  format %{
    $$template
    $$emit$$"lock addl [rsp + #0], 0\t! membar_volatile"
  %}
  ins_encode %{
    __ membar(Assembler::StoreLoad);
  %}
  ins_pipe(pipe_slow);
%}
```

**Comment by ashlee618** [comment]
Thanks for your answer, it was very helpful for me.

I checked again and found that `serverIps` and `syncConfigMap` are also similar to `leaderInfo`.

I feel that it would be better to continue to use `volatile` modification here.

```java
  private final class Connector extends WakeUpLoopRunnable {

    @Override
    public void runUnthrowable() {
      // here
      Set<String> ips = serverIps;
      try {
        tryConnectAllServer(ips);
      } catch (Throwable e) {
        LOGGER.error("failded to connect {}", ips, e);
      }
    }

    @Override
    public int getWaitingMillis() {
      return 3000;
    }
  }
```

```java
  public void refreshClusterInfos() {
    Set<MultiClusterSyncInfo> updates = multiClusterSyncRepository.queryLocalSyncInfos();
    Set<String> remoteDataCenters =
        updates.stream().map(MultiClusterSyncInfo::getRemoteDataCenter).collect(Collectors.toSet());
    // here
    Set<String> removes = Sets.difference(syncConfigMap.keySet(), remoteDataCenters);

    synchronized (this) {
      for (MultiClusterSyncInfo update : updates) {
        syncConfigMap.put(update.getRemoteDataCenter(), update);
        MULTI_CLUSTER_CONFIG_LOGGER.info("[refreshClusterInfos]update remote: {} config.", update);
      }

      for (String remove : removes) {
        syncConfigMap.remove(remove);
        removeLeader(remove);
        MULTI_CLUSTER_CONFIG_LOGGER.info("[refreshClusterInfos]remove remote: {} config.", remove);
      }
    }
  }

File: server/store/api/src/main/java/com/alipay/sofa/registry/store/api/elector/AbstractLeaderElector.java [hunk 1]
```diff
@@ -39,9 +39,9 @@ public abstract class AbstractLeaderElector implements LeaderElector {
 
   private volatile LeaderInfo leaderInfo = LeaderInfo.HAS_NO_LEADER;
 
-  private volatile boolean startElector = false;
```
Comment (lystander) [inline server/store/api/src/main/java/com/alipay/sofa/registry/store/api/elector/AbstractLeaderElector.java:42]: The get method of startElector is not synchronized, so I think the volatile keyword for this variable should not be removed.

Comment (ashlee618) [reply server/store/api/src/main/java/com/alipay/sofa/registry/store/api/elector/AbstractLeaderElector.java:42]: yes, my bad. startElector should not remove volatile.

fixed up.

**Review by lystander — APPROVED** [review]
LGTM

**Comment by NickNYU** [comment]
LGTM

**Comment by NickNYU** [comment]
@nobodyiam , please see if it could be merged

Thx :)

**Review by NickNYU — APPROVED** [review]

**Pull request merged**